src/ssl.c

```diff
@@ -62915,13 +62919,333 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
+ *
+ * pkcs7 - initialized PKCS7 structure, populated with signer, etc
+ * in    - input data
+ * flags - flags to control PKCS7 behavior. Other flags except those noted
+ *         below are ignored:
+ *
+ *    PKCS7_BINARY - Do not translate input data to MIME canonical
+ *                   format (\r\n line endings), thus preventing corruption of
+ *                   binary content.
+ *    PKCS7_TEXT   - Prepend MIME headers for text/plain to content.
+ *
+ * Returns 1 on success, 0 on error
+ */
+int wolfSSL_PKCS7_final(PKCS7* pkcs7, WOLFSSL_BIO* in, int flags)
+{
+    int ret = 1;
+    int memSz = 0;
+    unsigned char* mem = NULL;
+    WOLFSSL_PKCS7* p7 = (WOLFSSL_PKCS7*)pkcs7;
+    WOLFSSL_BIO* data = NULL;
+    static const char contTypeText[] = "Content-Type: text/plain\r\n\r\n";
+
+    WOLFSSL_ENTER("wolfSSL_PKCS7_final");
+
+    if (p7 == NULL || in == NULL) {
+        WOLFSSL_MSG("Bad input args to PKCS7_final");
+        ret = 0;
+    }
+
+    if (ret == 1) {
+        if ((data = wolfSSL_BIO_new(wolfSSL_BIO_s_mem())) == NULL) {
+            WOLFSSL_MSG("Error in wolfSSL_BIO_new");
+            ret = 0;
+        }
+    }
+
+    /* prepend Content-Type header if PKCS7_TEXT */
+    if ((ret == 1) && (flags & PKCS7_TEXT)) {
+        if (wolfSSL_BIO_write(data, contTypeText,
+                              (int)XSTRLEN(contTypeText)) < 0) {
+            WOLFSSL_MSG("Error prepending Content-Type header");
+            ret = 0;
+        }
+    }
+
+    /* convert line endings to CRLF if !PKCS7_BINARY */
+    if (ret == 1) {
+        if (flags & PKCS7_BINARY) {
+            /* no CRLF conversion, direct copy content */
+            if ((memSz = wolfSSL_BIO_get_mem_data(in, &mem)) < 0) {
+                ret = 0;
+            }
+            else if (wolfSSL_BIO_write(data, mem, memSz) < 0) {
+                ret = 0;
+            }
+        }
+        else {
+    #ifdef HAVE_SMIME
+            /* convert content line endings to CRLF */
+            if (wolfSSL_BIO_to_MIME_crlf(in, data) != 0) {
+                WOLFSSL_MSG("Error converting line endings to CRLF");
+                ret = 0;
+            }
+            else {
+                p7->pkcs7.contentCRLF = 1;
+            }
+    #else
+            WOLFSSL_MSG("Without PKCS7_BINARY requires wolfSSL to be built "
+                        "with HAVE_SMIME");
+            ret = 0;
+    #endif
+        }
+    }
+
+    if ((ret == 1) && ((memSz = wolfSSL_BIO_get_mem_data(data, &mem)) < 0)) {
+        WOLFSSL_MSG("Error in wolfSSL_BIO_get_mem_data");
+        ret = 0;
+    }
+
+    if (ret == 1) {
+        if (p7->data != NULL) {
+            XFREE(p7->data, NULL, DYNAMIC_TYPE_PKCS7);
+        }
+        p7->data = (byte*)XMALLOC(memSz, NULL, DYNAMIC_TYPE_PKCS7);
+        if (p7->data == NULL) {
+            ret = 0;
+        }
+        else {
+            XMEMCPY(p7->data, mem, memSz);
+            p7->len = memSz;
+        }
+    }
```

Comment:
Please use the following construct to extract the buffer (looking at our `BIO_free` we support `BIO_NOCLOSE` on a memory buffer so this should work):
```
 BUF_MEM *bptr;

 BIO_get_mem_ptr(mem, &bptr);
 BIO_set_close(mem, BIO_NOCLOSE); /* So BIO_free() leaves BUF_MEM alone */
 BIO_free(mem);
```
Taken from https://www.openssl.org/docs/manmaster/man3/BIO_get_mem_data.html

Reply:
Sure, I can switch from `BIO_get_mem_data()` to `BIO_get_mem_ptr()`.  Doesn't seem to save much though - about the same line count, but one extra struct pointer added to the function.

Reply:
Using the way I posted above saves on the `malloc` and `memcpy`. You take control of the internal memory buffer so you can just assign it to `p7->data`. Actually this might not be possible with how we have the mallocs set up. Using BIO_get_mem_ptr would make the heap and type hints to not line up with the ones that will be used to free them. Do we have a way to know if the heap and type hints are actually used? If we do then we could avoid the `malloc` and `memcpy` but otherwise it will be best to leave it how you wrote it previously.

Reply:
Actually please revert back to using `wolfSSL_BIO_get_mem_data`. Using `BIO_get_mem_ptr` won't save on a copy (as I originally thought it would) and with my memory BIO refactor in https://github.com/wolfSSL/wolfssl/pull/4956 it will be more correct to use `wolfSSL_BIO_get_mem_data`.

Reply:
Sounds good, reverted back.